packages/amplify_core/lib/src/category/amplify_storage_category.dart

```diff
@@ -25,55 +25,167 @@ part of amplify_interface;
 /// The Amplify CLI helps you to create and configure the storage category
 /// and auth category.
 /// {@endtemplate}
-class StorageCategory extends AmplifyCategory<StoragePluginInterface> {
+class StorageCategory<
+    PluginStorageListOperation extends StorageListOperation,
+    PluginStorageListOptions extends StorageListOptions,
+    PluginStorageGetPropertiesOperation extends StorageGetPropertiesOperation,
+    PluginStorageGetPropertiesOptions extends StorageGetPropertiesOptions,
+    PluginStorageGetUrlOperation extends StorageGetUrlOperation,
+    PluginStorageGetUrlOptions extends StorageGetUrlOptions,
+    PluginStorageUploadDataOperation extends StorageUploadDataOperation,
+    PluginStorageUploadDataOptions extends StorageUploadDataOptions,
+    PluginStorageCopyOperation extends StorageCopyOperation,
+    PluginStorageMoveOperation extends StorageMoveOperation,
+    PluginStorageRemoveOperation extends StorageRemoveOperation,
+    PluginStorageRemoveOptions extends StorageRemoveOptions,
+    PluginStorageRemoveManyOperation extends StorageRemoveManyOperation,
+    PluginStorageRemoveManyOptions extends StorageRemoveManyOptions,
+    PluginStorageItemWithAccessLevel extends StorageItemWithAccessLevel,
```

Comment:
I think breaking these down a bit more could help the APIs be expressed with the base types of the category. For example, right now `StorageItemWithAccessLevel` is a base type, but it's not clear why this should be the base type over `StorageItem` since all it does it add access level information to the storage item, and what value there is in allowing extension of StorageItemWithAccessLevel instead of or in addition to StorageItem. This is another way to do it:

```dart
class StorageCategory<
        PluginStorageItem extends StorageItem,
        ...> {

  PluginStorageCopyOperation copy({
    required StorageItemWithAccessLevel<PluginStorageItem> source,
    required StorageItemWithAccessLevel<PluginStorageItem> destination,
  }) {
    // ...
  }
}

class StorageItemWithAccessLevel<Item extends StorageItem> {
  /// {@macro amplify_core.storage.storage_item_access_level}
  const StorageItemWithAccessLevel({
    required this.storageItem,
    required this.storageAccessLevel,
  });

  final Item storageItem;
  final StorageAccessLevel storageAccessLevel;
}

class StorageCopyRequest<Item extends StorageItem> {
  /// {@macro amplify_core.storage.copy_request}
  const StorageCopyRequest({
    required this.source,
    required this.destination,
  });

  /// Copy source.
  final StorageItemWithAccessLevel<Item> source;

  /// Copy destination.
  final StorageItemWithAccessLevel<Item> destination;
}
```

Now, everything is expressed in terms of `StorageItem` and any extension should be made to that. You can additionally allow extending `StorageItemWithAccessLevel` if you want.

Or should `accessLevel` be a property on `StorageItem`? I don't see the value in having two separate types if `StorageItem` doesn't show up anywhere in the public API.

Reply:
Nice, good suggestion.